File: lib.py/treebuilders/simpletree.py
#!/usr/bin/env python
import _base
from xml.sax.saxutils import escape

# DOM-core like implementation with extensions.
class Node(_base.Node):
    def __init__(self):
        self.childNodes = []

    def __iter__(self):
        for node in self.childNodes:
            yield node
            for item in node:
                yield item

    def __unicode__(self):
        return self.name

    def toxml(self):
        raise NotImplementedError

    def printTree(self, indent=0):
        tree = '\n|%s%s' % (' '* indent, unicode(self))
        for child in self.childNodes:
            tree += child.printTree(indent + 2)
        return tree

    def appendChild(self, node, index=None):
        if (isinstance(node, Text) and self.childNodes and
          isinstance(self.childNodes[-1], Text)):
            self.childNodes[-1].value += node.value
        else:
            self.childNodes.append(node)
        node.parent = self

    def insertText(self, data):
        self.appendChild(Text(data))

    def hasContent(self):
        """Return true if the node has children or text"""
        return bool(self.childNodes)

class Document(Node):
    def __init__(self):
        Node.__init__(self)

    def __unicode__(self):
        return "#document"

    def toxml(self, encoding="utf=8"):
        result = ""
        for child in self.childNodes:
            result += child.toxml()
        return result.encode(encoding)

    def hilite(self, encoding="utf-8"):
        result = "<pre>"
        for child in self.childNodes:
            result += child.hilite()
        return result.encode(encoding) + "</pre>"
    
    def printTree(self):
        tree = unicode(self)
        for child in self.childNodes:
            tree += child.printTree(1)
        return tree

class Text(Node):
    def __init__(self, value):
        Node.__init__(self)
        self.value = value

    def __unicode__(self):
        return "\"%s\"" % self.value

    def toxml(self):
        return escape(self.value)

    hilite = toxml

class Element(Node):
    def __init__(self, name, prefix, localname, namespace, attributes):
        Node.__init__(self)
        self.name = name
        self.prefix = prefix
        self.localname = localname
        self.namespace = namespace
        self.attributes = attributes

    def __unicode__(self):
        return "<%s>" % self.name

    def printTree(self, indent):
        tree = '\n|%s<%s> (%s, %s, %s)' % (' '*indent, self.name, self.prefix, self.localname, self.namespace)
        indent += 2
        if self.attributes:
            for token in self.attributes:
                tree += '\n|%s%s="%s" (%s, %s, %s)' % (' ' * indent, token["name"], token["value"], token["prefix"], token["localname"], token["namespace"])
        for child in self.childNodes:
            tree += child.printTree(indent)
        return tree

class Pi(Node):
    def __init__(self, name, data):
        Node.__init__(self)
        self.name = name
        self.data = data
    
    def __unicode__(self):
        return "<?%s %s?>" % (self.name, self.data)

    def toxml(self):
        return "<?%s %s?>" % (self.name, self.data)

    def hilite(self):
        return '<code class="markup pi">&lt;?%s %s?></code>' % (escape(self.name), escape(self.data))

class Comment(Node):
    def __init__(self, data):
        Node.__init__(self)
        self.data = data

    def __unicode__(self):
        return "<!-- %s -->" % self.data
    
    def toxml(self):
        return "<!--%s-->" % self.data

    def hilite(self):
        return '<code class="markup comment">&lt;!--%s--></code>' % escape(self.data)

class TreeBuilder(_base.TreeBuilder):
    documentClass = Document
    elementClass = Element
    piClass = Pi
    commentClass = Comment

    def testSerializer(self, node):
        return node.printTree()

File: lib.rb/treebuilders/simpletree.rb
require 'treebuilders/base'

# DOM-core like implementation with extensions.
module SimpleTree
  class Node < Base::Node
    def initialize
      @childNodes = []
    end

    def each
      for node in @childNodes
        yield node
        for item in node
          yield item
        end
      end
    end

    def to_s
      return @name
    end

    def to_xml
      raise NotImplementedError
    end

    def printTree(indent=0)
      tree = "\n|#{' ' * indent}#{to_s}"
      for child in @childNodes
        tree += child.printTree(indent + 2)
      end
      return tree
    end

    def appendChild(node, index=nil)
      if (node.instance_of?(Text) and not @childNodes.empty? and
        @childNodes.last.instance_of?(Text))
        @childNodes.last.value += node.value
      else
        @childNodes << node
      end
      node.parent = self
    end

    def insertText(data)
      appendChild(Text.new(data))
    end

    def hasContent
      """Return true if the node has children or text"""
      return bool(@childNodes)
    end
  end

  class Document < Node
    def initialize
      super
    end

    def to_s
      return "#document"
    end

    def to_xml(encoding="utf=8")
      result = ""
      for child in @childNodes
        result += child.to_xml()
      end
      return result.encode(encoding)
    end

    def hilite(encoding="utf-8")
      result = "<pre>"
      for child in @childNodes
        result += child.hilite()
      end
      return result.encode(encoding) + "</pre>"
    end

    def printTree
      tree = self.to_s
      for child in @childNodes
        tree += child.printTree(1)
      end
      return tree
    end
  end

  class Text < Node
    attr_accessor :value

    def initialize(value)
      super()
      @value = value
    end

    def to_s
      return "\"%s\"" % @value
    end

    def to_xml
      return escape(@value)
    end

    alias :hilite :to_xml
  end

  class Element < Node
    attr_reader :name
    attr_reader :attributes

    def initialize(name, prefix, localname, namespace, attributes)
      super()
      @name = name
      @prefix = prefix
      @localname = localname
      @namespace = namespace
      @attributes = attributes
    end

    def to_s
      return "<%s>" % @name
    end

    def printTree(indent)
      tree = "\n|#{' ' * indent}<#{@name}> " +
        "(#{@prefix}, #{@localname}, #{@namespace})"
      indent += 2
      if @attributes
        for token in @attributes
          tree += "\n|#{' ' * indent}#{token["name"]}=\"#{token["value"]}\" " +
            "(#{token["prefix"]}, #{token["localname"]}, #{token["namespace"]})"
        end
      end
      for child in @childNodes
        tree += child.printTree(indent)
      end
      return tree
    end
  end

  class Pi < Node
    def initialize(name, data)
      super()
      @name = name
      @data = data
    end

    def to_s
      return "<?#{@name} #{@data}?>"
    end

    def to_xml
      return "<?#{@name} #{@data}?>"
    end

    def hilite
      return '<code class="markup pi">&lt;?%s %s?></code>' % [escape(@name),
      escape(@data)]
    end
  end

  class Comment < Node
    def initialize(data)
      super()
      @data = data
    end

    def to_s
      return "<!-- %s -->" % @data
    end

    def to_xml
      return "<!--%s-->" % @data
    end

    def hilite
      return '<code class="markup comment">&lt;!--%s--></code>' % escape(@data)
    end
  end

  class TreeBuilder < Base::TreeBuilder
    @@documentClass = Document
    @@elementClass = Element
    @@piClass = Pi
    @@commentClass = Comment

    def testSerializer(node)
      return node.printTree()
    end
  end
end
